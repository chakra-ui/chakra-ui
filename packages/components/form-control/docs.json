{
  "FormControl": {
    "isDisabled": {
      "type": "boolean",
      "defaultValue": false,
      "required": false,
      "description": "If `true`, the form control will be disabled. This has 2 side effects:\n- The `FormLabel` will have `data-disabled` attribute\n- The form element (e.g, Input) will be disabled"
    },
    "isInvalid": {
      "type": "boolean",
      "defaultValue": false,
      "required": false,
      "description": "If `true`, the form control will be invalid. This has 2 side effects:\n- The `FormLabel` and `FormErrorIcon` will have `data-invalid` set to `true`\n- The form element (e.g, Input) will have `aria-invalid` set to `true`"
    },
    "isReadOnly": {
      "type": "boolean",
      "defaultValue": false,
      "required": false,
      "description": "If `true`, the form control will be readonly"
    },
    "isRequired": {
      "type": "boolean",
      "defaultValue": false,
      "required": false,
      "description": "If `true`, the form control will be required. This has 2 side effects:\n- The `FormLabel` will show a required indicator\n- The form element (e.g, Input) will have `aria-required` set to `true`"
    },
    "label": {
      "type": "string",
      "required": false,
      "description": "The label text used to inform users as to what information is\nrequested for a text field."
    }
  },
  "FormHelperText": {},
  "UseFormControl": {
    "aria-describedby": { "type": "string", "required": false },
    "disabled": { "type": "boolean", "required": false },
    "id": { "type": "string", "required": false },
    "isDisabled": {
      "type": "boolean",
      "defaultValue": false,
      "required": false,
      "description": "If `true`, the form control will be disabled. This has 2 side effects:\n- The `FormLabel` will have `data-disabled` attribute\n- The form element (e.g, Input) will be disabled"
    },
    "isInvalid": {
      "type": "boolean",
      "defaultValue": false,
      "required": false,
      "description": "If `true`, the form control will be invalid. This has 2 side effects:\n- The `FormLabel` and `FormErrorIcon` will have `data-invalid` set to `true`\n- The form element (e.g, Input) will have `aria-invalid` set to `true`"
    },
    "isReadOnly": {
      "type": "boolean",
      "defaultValue": false,
      "required": false,
      "description": "If `true`, the form control will be readonly"
    },
    "isRequired": {
      "type": "boolean",
      "defaultValue": false,
      "required": false,
      "description": "If `true`, the form control will be required. This has 2 side effects:\n- The `FormLabel` will show a required indicator\n- The form element (e.g, Input) will have `aria-required` set to `true`"
    },
    "onBlur": { "type": "FocusEventHandler<T>", "required": false },
    "onFocus": { "type": "FocusEventHandler<T>", "required": false },
    "readOnly": { "type": "boolean", "required": false },
    "required": { "type": "boolean", "required": false }
  },
  "FormErrorMessage": {},
  "FormLabel": {
    "optionalIndicator": {
      "type": "type ONLY_FOR_FORMAT =\n  | string\n  | number\n  | boolean\n  | ReactElement<any, string | JSXElementConstructor<any>>\n  | ReactFragment\n  | ReactPortal",
      "required": false
    },
    "requiredIndicator": {
      "type": "ReactElement<any, string | JSXElementConstructor<any>>",
      "required": false
    }
  },
  "RequiredIndicator": {}
}
