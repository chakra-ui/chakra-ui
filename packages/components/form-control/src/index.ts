export {
  FormControl,
  FormHelperText,
  useFormControlContext,
  useFormControlStyles,
} from "./form-control"
export type {
  FormControlOptions,
  FormControlProps,
  FormHelperTextProps,
} from "./form-control"
export { useFormControl, useFormControlProps } from "./use-form-control"
export type { UseFormControlProps } from "./use-form-control"
export {
  FormErrorIcon,
  FormErrorMessage,
  useFormErrorStyles,
} from "./form-error"
export type { FormErrorMessageProps } from "./form-error"
export { FormLabel } from "./form-label"
export type { FormLabelProps } from "./form-label"
