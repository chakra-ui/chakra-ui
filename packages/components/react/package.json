{
  "name": "@chakra-ui/react",
  "version": "2.4.9",
  "description": "Responsive and accessible React UI components built with React and Emotion",
  "main": "src/index.ts",
  "files": [
    "dist"
  ],
  "sideEffects": false,
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@chakra-ui/accordion": "workspace:*",
    "@chakra-ui/alert": "workspace:*",
    "@chakra-ui/avatar": "workspace:*",
    "@chakra-ui/breadcrumb": "workspace:*",
    "@chakra-ui/button": "workspace:*",
    "@chakra-ui/card": "workspace:*",
    "@chakra-ui/checkbox": "workspace:*",
    "@chakra-ui/close-button": "workspace:*",
    "@chakra-ui/control-box": "workspace:*",
    "@chakra-ui/counter": "workspace:*",
    "@chakra-ui/css-reset": "workspace:*",
    "@chakra-ui/editable": "workspace:*",
    "@chakra-ui/form-control": "workspace:*",
    "@chakra-ui/hooks": "workspace:*",
    "@chakra-ui/icon": "workspace:*",
    "@chakra-ui/image": "workspace:*",
    "@chakra-ui/input": "workspace:*",
    "@chakra-ui/layout": "workspace:*",
    "@chakra-ui/live-region": "workspace:*",
    "@chakra-ui/media-query": "workspace:*",
    "@chakra-ui/menu": "workspace:*",
    "@chakra-ui/modal": "workspace:*",
    "@chakra-ui/number-input": "workspace:*",
    "@chakra-ui/pin-input": "workspace:*",
    "@chakra-ui/popover": "workspace:*",
    "@chakra-ui/popper": "workspace:*",
    "@chakra-ui/portal": "workspace:*",
    "@chakra-ui/progress": "workspace:*",
    "@chakra-ui/provider": "workspace:*",
    "@chakra-ui/radio": "workspace:*",
    "@chakra-ui/react-env": "workspace:*",
    "@chakra-ui/select": "workspace:*",
    "@chakra-ui/skeleton": "workspace:*",
    "@chakra-ui/slider": "workspace:*",
    "@chakra-ui/spinner": "workspace:*",
    "@chakra-ui/stat": "workspace:*",
    "@chakra-ui/styled-system": "workspace:*",
    "@chakra-ui/switch": "workspace:*",
    "@chakra-ui/system": "workspace:*",
    "@chakra-ui/table": "workspace:*",
    "@chakra-ui/tabs": "workspace:*",
    "@chakra-ui/tag": "workspace:*",
    "@chakra-ui/textarea": "workspace:*",
    "@chakra-ui/theme": "workspace:*",
    "@chakra-ui/theme-utils": "workspace:*",
    "@chakra-ui/toast": "workspace:*",
    "@chakra-ui/tooltip": "workspace:*",
    "@chakra-ui/transition": "workspace:*",
    "@chakra-ui/utils": "workspace:*",
    "@chakra-ui/visually-hidden": "workspace:*"
  },
  "peerDependencies": {
    "@emotion/react": "^11.0.0",
    "@emotion/styled": "^11.0.0",
    "framer-motion": ">=4.0.0",
    "react": ">=18",
    "react-dom": ">=18"
  },
  "author": "Segun Adebayo <sage@adebayosegun.com>",
  "license": "MIT",
  "homepage": "https://chakra-ui.com/",
  "repository": {
    "type": "git",
    "url": "https://github.com/chakra-ui/chakra-ui",
    "directory": "packages/components/react"
  },
  "keywords": [
    "react",
    "ui",
    "design-system",
    "react-components",
    "uikit",
    "accessible",
    "components",
    "emotion",
    "library",
    "design-system"
  ],
  "storybook": {
    "title": "Chakra UI",
    "url": "https://storybook.chakra-ui.com"
  },
  "devDependencies": {
    "@emotion/react": "^11.9.0",
    "@emotion/styled": "^11.8.1",
    "framer-motion": "^6.2.9",
    "react": "^18.0.0",
    "react-dom": "^18.0.0",
    "clean-package": "2.1.1"
  },
  "scripts": {
    "build": "tsup src --dts",
    "dev": "pnpm build:fast -- --watch",
    "clean": "rimraf dist .turbo",
    "typecheck": "tsc --noEmit",
    "build:fast": "tsup src",
    "prepack": "clean-package",
    "postpack": "clean-package restore"
  },
  "clean-package": "../../../clean-package.config.json",
  "tsup": {
    "clean": true,
    "target": "es2019",
    "format": [
      "cjs",
      "esm"
    ]
  }
}
