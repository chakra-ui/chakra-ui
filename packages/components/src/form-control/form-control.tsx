import { omitThemingProps, ThemingProps } from "@chakra-ui/styled-system"
import { cx } from "@chakra-ui/utils/cx"
import {
  chakra,
  forwardRef,
  HTMLChakraProps,
  useMultiStyleConfig,
} from "../system"
import {
  FormControlProvider,
  FormControlStylesProvider,
} from "./form-control-context"
import { FormControlContext } from "./types"
import { useFormControlProvider } from "./use-form-control-provider"

export interface FormControlProps
  extends HTMLChakraProps<"div">,
    ThemingProps<"FormControl">,
    FormControlContext {}

/**
 * FormControl provides context such as
 * `isInvalid`, `isDisabled`, and `isRequired` to form elements.
 *
 * This is commonly used in form elements such as `input`,
 * `select`, `textarea`, etc.
 *
 * @see Docs https://chakra-ui.com/docs/components/form-control
 */
export const FormControl = forwardRef<FormControlProps, "div">(
  function FormControl(props, ref) {
    const styles = useMultiStyleConfig("Form", props)
    const ownProps = omitThemingProps(props)
    const { getRootProps, ...context } = useFormControlProvider(ownProps)

    const className = cx("chakra-form-control", props.className)

    return (
      <FormControlProvider value={context}>
        <FormControlStylesProvider value={styles}>
          <chakra.div
            {...getRootProps({}, ref)}
            className={className}
            __css={styles["container"]}
          />
        </FormControlStylesProvider>
      </FormControlProvider>
    )
  },
)

FormControl.displayName = "FormControl"
