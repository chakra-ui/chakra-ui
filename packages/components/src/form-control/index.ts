export {
  FormControl,
  FormHelperText,
  useFormControlContext,
  useFormControlStyles,
} from "./form-control"
export type {
  FormControlOptions,
  FormControlProps,
  FormHelperTextProps,
} from "./form-control"
export { useFormControl, useFormControlProps } from "./use-form-control"
export type { UseFormControlProps } from "./use-form-control"
export {
  FormErrorIcon,
  FormErrorMessage,
  useFormErrorStyles,
} from "./form-error"
export type { FormErrorMessageProps } from "./form-error"
export { FormLabel, RequiredIndicator } from "./form-label"
export type { FormLabelProps, RequiredIndicatorProps } from "./form-label"
