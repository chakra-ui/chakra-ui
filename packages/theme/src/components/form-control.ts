import { formAnatomy as parts } from "@chakra-ui/anatomy"
import {
  createMultiStyleConfigHelpers,
  cssVar,
  defineStyle,
} from "@chakra-ui/styled-system"

const { definePartsStyle, defineMultiStyleConfig } =
  createMultiStyleConfigHelpers(parts.keys)

const $fg = cssVar("form-control-color")

const baseStyleRequiredIndicator = defineStyle({
  marginStart: "1",
  [$fg.variable]: "colors.red.500",
  _dark: {
    [$fg.variable]: "colors.red.300",
  },
  color: $fg.reference,
})

const baseStyleHelperText = defineStyle({
  mt: "2",
  [$fg.variable]: "colors.gray.600",
  _dark: {
    [$fg.variable]: "colors.whiteAlpha.600",
  },
  color: $fg.reference,
  lineHeight: "normal",
  fontSize: "sm",
})

const baseStyle = definePartsStyle({
  container: {
    width: "100%",
    position: "relative",
  },
  requiredIndicator: baseStyleRequiredIndicator,
  helperText: baseStyleHelperText,
})

export const formTheme = defineMultiStyleConfig({
  baseStyle,
})
