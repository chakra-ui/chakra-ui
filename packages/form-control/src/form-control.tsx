import { useBoolean, useId, useSafeLayoutEffect } from "@chakra-ui/hooks"
import {
  chakra,
  forwardRef,
  omitThemingProps,
  StylesProvider,
  ThemingProps,
  useMultiStyleConfig,
  useStyles,
  HTMLChakraProps,
} from "@chakra-ui/system"
import { createContext, cx, __DEV__ } from "@chakra-ui/utils"
import * as React from "react"

export interface FormControlOptions {
  /**
   * If `true`, the form control will be required. This has 2 side effects:
   * - The `FormLabel` will show a required indicator
   * - The form element (e.g, Input) will have `aria-required` set to `true`
   */
  isRequired?: boolean
  /**
   * If `true`, the form control will be disabled. This has 2 side effects:
   * - The `FormLabel` will have `data-disabled` attribute
   * - The form element (e.g, Input) will be disabled
   */
  isDisabled?: boolean
  /**
   * If `true`, the form control will be invalid. This has 2 side effects:
   * - The `FormLabel` and `FormErrorIcon` will have `data-invalid` set to `true`
   * - The form element (e.g, Input) will have `aria-invalid` set to `true`
   */
  isInvalid?: boolean
  /**
   * If `true`, the form control will be readonly
   */
  isReadOnly?: boolean
}

interface FormControlContext extends FormControlOptions {
  /**
   * The label text used to inform users as to what information is
   * requested for a text field.
   */
  label?: string
  /**
   * The custom `id` to use for the form control. This is passed directly to the form element (e.g, Input).
   * - The form element (e.g Input) gets the `id`
   * - The form label id: `form-label-${id}`
   * - The form error text id: `form-error-text-${id}`
   * - The form helper text id: `form-helper-text-${id}`
   */
  id?: string
}

type ControlContext = Omit<
  ReturnType<typeof useFormControlProvider>,
  "htmlProps"
>

const [
  FormControlProvider,
  useFormControlContext,
] = createContext<ControlContext>({
  strict: false,
  name: "FormControlContext",
})

export { useFormControlContext }

function useFormControlProvider(props: FormControlContext) {
  const {
    id: idProp,
    isRequired,
    isInvalid,
    isDisabled,
    isReadOnly,
    ...htmlProps
  } = props

  // Generate all the required ids
  const uuid = useId()
  const id = idProp || `field-${uuid}`

  const labelId = `${id}-label`
  const feedbackId = `${id}-feedback`
  const helpTextId = `${id}-helptext`

  /**
   * Track whether the `FormErrorMessage` has been rendered.
   * We use this to append its id the the `aria-describedby` of the `input`.
   */
  const [hasFeedbackText, setHasFeedbackText] = useBoolean()

  /**
   * Track whether the `FormHelperText` has been rendered.
   * We use this to append its id the the `aria-describedby` of the `input`.
   */
  const [hasHelpText, setHasHelpText] = useBoolean()

  // Track whether the form element (e.g, `input`) has focus.
  const [isFocused, setFocus] = useBoolean()

  const context = {
    isRequired: !!isRequired,
    isInvalid: !!isInvalid,
    isReadOnly: !!isReadOnly,
    isDisabled: !!isDisabled,
    isFocused: !!isFocused,
    onFocus: setFocus.on,
    onBlur: setFocus.off,
    hasFeedbackText,
    setHasFeedbackText,
    hasHelpText,
    setHasHelpText,
    id,
    labelId,
    feedbackId,
    helpTextId,
    htmlProps,
  }

  return context
}

export interface FormControlProps
  extends HTMLChakraProps<"div">,
    ThemingProps<"FormControl">,
    FormControlContext {}

/**
 * FormControl provides context such as
 * `isInvalid`, `isDisabled`, and `isRequired` to form elements.
 *
 * This is commonly used in form elements such as `input`,
 * `select`, `textarea`, etc.
 */
export const FormControl = forwardRef<FormControlProps, "div">((props, ref) => {
  const styles = useMultiStyleConfig("Form", props)
  const ownProps = omitThemingProps(props)
  const { htmlProps, ...context } = useFormControlProvider(ownProps)

  const _className = cx("chakra-form-control", props.className)

  return (
    <FormControlProvider value={context}>
      <StylesProvider value={styles}>
        <chakra.div
          role="group"
          ref={ref}
          {...htmlProps}
          className={_className}
          __css={{
            width: "100%",
            position: "relative",
          }}
        />
      </StylesProvider>
    </FormControlProvider>
  )
})

if (__DEV__) {
  FormControl.displayName = "FormControl"
}

export interface HelpTextProps extends HTMLChakraProps<"div"> {}

/**
 * FormHelperText
 *
 * Assistive component that conveys additional guidance
 * about the field, such as how it will be used and what
 * types in values should be provided.
 */
export const FormHelperText = forwardRef<HelpTextProps, "div">((props, ref) => {
  const field = useFormControlContext()
  const styles = useStyles()

  /**
   * Notify the field context when the help text is rendered on screen,
   * so we can apply the correct `aria-describedby` to the field (e.g. input, textarea).
   */
  useSafeLayoutEffect(() => {
    field?.setHasHelpText.on()
    return () => field?.setHasHelpText.off()
  }, [])

  const _className = cx("chakra-form__helper-text", props.className)

  return (
    <chakra.div
      ref={ref}
      __css={styles.helperText}
      {...props}
      className={_className}
      id={props.id ?? field?.helpTextId}
    />
  )
})

if (__DEV__) {
  FormHelperText.displayName = "FormHelperText"
}
