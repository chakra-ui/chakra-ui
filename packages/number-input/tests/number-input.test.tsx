/* eslint-disable jsx-a11y/label-has-associated-control */
import { FormControl, FormHelperText, FormLabel } from "@chakra-ui/form-control"
import {
  testA11y,
  fireEvent,
  render,
  renderHook,
  userEvent,
  press,
  screen,
} from "@chakra-ui/test-utils"
import * as React from "react"
import {
  NumberDecrementStepper,
  NumberIncrementStepper,
  NumberInput,
  NumberInputField,
  NumberInputProps,
  NumberInputStepper,
  useNumberInput,
} from "../src"

function renderComponent(props: NumberInputProps = {}) {
  return render(
    <>
      <label htmlFor="input">Select number:</label>
      <NumberInput id="input" data-testid="root" {...props}>
        <NumberInputField data-testid="input" />
        <NumberInputStepper data-testid="group">
          <NumberIncrementStepper children="+" data-testid="up-btn" />
          <NumberDecrementStepper children="-" data-testid="down-btn" />
        </NumberInputStepper>
      </NumberInput>
    </>,
  )
}

/**
 * Get some inspiration here
 * https://github.com/palantir/blueprint/blob/3aa56473d253f5287e0960759bee367a9ff3e045/packages/core/test/controls/numericInputTests.tsx
 * https://github.com/deberoppa7/react-numeric-input/blob/master/src/index.test.js
 */

test("passes a11y test", async () => {
  const { container } = renderComponent()
  await testA11y(container)
})

test("should start with empty string", () => {
  const { result } = renderHook(() => useNumberInput())
  expect(result.current.value).toBe("")
})

test("should increment on press increment button", () => {
  const { getByTestId } = renderComponent()

  const upBtn = getByTestId("up-btn")
  const input = getByTestId("input")

  fireEvent.mouseDown(upBtn)
  // since the input's value is empty, this will set it to `step`
  // which is `1` by default
  expect(input).toHaveValue("1")

  fireEvent.mouseDown(upBtn)
  expect(input).toHaveValue("2")
})

test("should increase/decrease with keyboard", () => {
  const { getByTestId } = renderComponent()

  const input = getByTestId("input")

  input.focus()

  press.ArrowUp(input)
  press.ArrowUp(input)
  press.ArrowUp(input)
  expect(input).toHaveValue("3")

  press.ArrowDown(input)
  press.ArrowDown(input)
  press.ArrowDown(input)
  expect(input).toHaveValue("0")

  press.ArrowUp(input)
  expect(input).toHaveValue("1")

  press.Home(input)
  expect(input).toHaveValue("-9007199254740991")

  press.End(input)
  expect(input).toHaveValue("9007199254740991")
})

test("should increase/decrease by 10*step on shift+Arrow", () => {
  const { getByTestId } = renderComponent({ defaultValue: 0 })

  const input = getByTestId("input")

  press.ArrowUp(input)
  expect(input).toHaveValue("1")
  press.ArrowUp(input, { shiftKey: true })
  expect(input).toHaveValue("11")

  press.ArrowDown(input, { shiftKey: true })
  expect(input).toHaveValue("1")
  press.ArrowDown(input)
  expect(input).toHaveValue("0")
})

test("should increase/decrease by 0.1*step on ctrl+Arrow", () => {
  const { getByTestId } = renderComponent({
    defaultValue: 0,
    step: 0.1,
    precision: 2,
  })

  const input = getByTestId("input")

  press.ArrowUp(input)
  expect(input).toHaveValue("0.10")
  press.ArrowUp(input, { ctrlKey: true })
  expect(input).toHaveValue("0.11")

  press.ArrowDown(input, { ctrlKey: true })
  expect(input).toHaveValue("0.10")
  press.ArrowDown(input)
  expect(input).toHaveValue("0.00")
})

it("should behave properly with precision value", () => {
  const { getByTestId } = renderComponent({
    defaultValue: 0,
    step: 0.65,
    precision: 2,
  })

  const input = getByTestId("input")
  const incBtn = getByTestId("up-btn")
  const decBtn = getByTestId("down-btn")

  expect(input).toHaveValue("0.00")
  userEvent.click(incBtn)
  expect(input).toHaveValue("0.65")
  userEvent.click(incBtn)
  expect(input).toHaveValue("1.30")
  userEvent.click(incBtn)
  expect(input).toHaveValue("1.95")
  userEvent.click(decBtn)
  expect(input).toHaveValue("1.30")

  // on blur, value is clamped using precision
  userEvent.type(input, "1234")
  expect(input).toHaveValue("1.301234")
  fireEvent.blur(input)
  expect(input).toHaveValue("1.30")
})

test("should call onChange on value change", () => {
  const onChange = jest.fn()
  const { getByTestId } = renderComponent({ onChange })

  const upBtn = getByTestId("up-btn")

  userEvent.click(upBtn)

  expect(onChange).toBeCalled()
  expect(onChange).toBeCalledWith("1", 1)
})

test("should constrain value onBlur", () => {
  const { getByTestId } = renderComponent({ max: 30 })

  const input = getByTestId("input")

  userEvent.type(input, "34.55")

  // value is beyond max so it should reset to `max`
  fireEvent.blur(input)

  expect(input).toHaveValue("30.00")
})

test("should focus input on spin", () => {
  const { getByTestId } = renderComponent()

  const input = getByTestId("input")
  const upBtn = getByTestId("up-btn")

  fireEvent.mouseDown(upBtn)
  expect(input).toHaveValue("1")

  // for some reason, .toHaveFocus assertion doesn't work
  // expect(tools.getByTestId("input")).toEqual(document.activeElement)
})

test("should derive values from surrounding FormControl", () => {
  const onFocus = jest.fn()
  const onBlur = jest.fn()

  render(
    <FormControl
      id="input"
      isRequired
      isInvalid
      isDisabled
      isReadOnly
      onFocus={onFocus}
      onBlur={onBlur}
    >
      <FormLabel>Number</FormLabel>
      <NumberInput data-testid="root">
        <NumberInputField data-testid="input" />
        <NumberInputStepper data-testid="group">
          <NumberIncrementStepper children="+" data-testid="up-btn" />
          <NumberDecrementStepper children="-" data-testid="down-btn" />
        </NumberInputStepper>
      </NumberInput>
      <FormHelperText>Select a number</FormHelperText>
    </FormControl>,
  )

  const input = screen.getByTestId("input")

  expect(input).toHaveAttribute("id", "input")
  expect(input).toHaveAttribute("aria-invalid", "true")
  expect(input).toHaveAttribute("aria-required", "true")
  expect(input).toHaveAttribute("aria-readonly", "true")
  expect(input).toHaveAttribute("aria-invalid", "true")
  expect(input).toHaveAttribute("aria-describedby")

  fireEvent.focus(input)
  expect(onFocus).toHaveBeenCalled()

  fireEvent.blur(input)
  expect(onBlur).toHaveBeenCalled()
})

test("should fallback to min if `e` is typed", () => {
  const { getByTestId } = renderComponent({ max: 30, min: 1 })
  const input = getByTestId("input")
  userEvent.type(input, "e")
  // value is beyond max so it should reset to `max`
  fireEvent.blur(input)
  expect(input).toHaveValue("1")
})
