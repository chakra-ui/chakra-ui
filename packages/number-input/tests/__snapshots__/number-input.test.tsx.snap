// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<DocumentFragment>
  <label
    for="input"
  >
    Select number:
  </label>
  .emotion-4 {
  position: relative;
}

.emotion-0 {
  width: 100%;
  outline: 0;
  position: relative;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  font-size: 1rem;
  padding-left: 1rem;
  padding-right: 1rem;
  height: 2.5rem;
  border-radius: 0.25rem;
  border: 1px solid;
  border-color: inherit;
  background: #FFFFFF;
}

.emotion-0:hover,
.emotion-0[data-hover] {
  border-color: #CBD5E0;
}

.emotion-0[aria-readonly=true],
.emotion-0[readonly],
.emotion-0[data-readonly] {
  box-shadow: none !important;
  -webkit-user-select: all;
  -moz-user-select: all;
  -ms-user-select: all;
  user-select: all;
}

.emotion-0[disabled],
.emotion-0[disabled]:focus,
.emotion-0[disabled]:hover,
.emotion-0[aria-disabled=true],
.emotion-0[aria-disabled=true]:focus,
.emotion-0[aria-disabled=true]:hover,
.emotion-0[data-disabled],
.emotion-0[data-disabled]:focus,
.emotion-0[data-disabled]:hover {
  opacity: 0.4;
  cursor: not-allowed;
}

.emotion-0:focus,
.emotion-0[data-focus] {
  z-index: 1;
  border-color: #3182CE;
  box-shadow: 0 0 0 1px #3182CE;
}

.emotion-0[aria-invalid=true],
.emotion-0[data-invalid] {
  border-color: #E53E3E;
  box-shadow: 0 0 0 1px #E53E3E;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: absolute;
  top: 0;
  right: 0px;
  margin: 1px;
  height: calc(100% - 2px);
  z-index: 1;
  width: 24px;
}

.emotion-1 {
  border-left: 1px solid;
  border-color: inherit;
  color: inherit;
  font-size: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  line-height: normal;
}

.emotion-1:active,
.emotion-1[data-active] {
  background: #E2E8F0;
}

.emotion-1[disabled],
.emotion-1[disabled]:focus,
.emotion-1[disabled]:hover,
.emotion-1[aria-disabled=true],
.emotion-1[aria-disabled=true]:focus,
.emotion-1[aria-disabled=true]:hover,
.emotion-1[data-disabled],
.emotion-1[data-disabled]:focus,
.emotion-1[data-disabled]:hover {
  opacity: 0.4;
  cursor: not-allowed;
}

.emotion-1:first-of-type {
  border-top-right-radius: 0.25rem;
}

.emotion-1:last-of-type {
  border-bottom-right-radius: 0.25rem;
  margin-top: -1px;
  border-top-width: 1px;
}

<div
    class="emotion-4"
    data-testid="root"
  >
    <input
      aria-invalid="false"
      aria-valuemax="9007199254740991"
      aria-valuemin="-9007199254740991"
      autocomplete="off"
      autocorrect="off"
      class="emotion-0"
      data-testid="input"
      id="input"
      inputmode="numeric"
      pattern="[0-9]*"
      role="spinbutton"
      type="text"
      value=""
    />
    <div
      aria-hidden="true"
      class="emotion-3"
      data-testid="group"
    >
      <div
        class="emotion-1"
        data-testid="up-btn"
        role="button"
        tabindex="-1"
      >
        +
      </div>
      <div
        class="emotion-1"
        data-testid="down-btn"
        role="button"
        tabindex="-1"
      >
        -
      </div>
    </div>
  </div>
</DocumentFragment>
`;
